### PR TITLE
Email type [WIP]

### PR DESCRIPTION
**Pull request by tlrobinson**
Adds an "Email" type which we format with "mailto:" links on the frontend.

We should also attempt to detect the Email type when syncing, like we do for URL.

(this builds on #3711)

### CODE REVIEW

**Comment by tlrobinson** [comment]
@camsaul I looked at adding email detection to sync and it looks like there's a bunch of code for doing it for URLs (`field-percent-urls`) that possibly could be refactored to match any pattern, like email, phone numbers, etc.

**Comment by salsakran** [comment]
front-end looks good to me 👍

**Comment by salsakran** [comment]
@camsaul I took a crack at the analyze addition for this. Can you review commit https://github.com/metabase/metabase/pull/3717/commits/212f47ffdfe8c352eeee54892736390e09ba56c4 , give me style feedback, etc?

File: src/metabase/sync_database/analyze.clj [hunk 1]
```diff
@@ -134,6 +134,45 @@
         (log/debug (u/format-color 'green "Field '%s' looks like it contains valid JSON objects. Setting special_type to :type/JSON." (field/qualified-name field)))
         (assoc field-stats :special-type :type/JSON, :preview-display false)))))
 
+(defn valid-email?
+  [email]
+  (let [pattern #"[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?"]
+    (and (string? email) (re-matches pattern email))))
```
Comment (camsaul) [inline src/metabase/sync_database/analyze.clj]: we already have a function called `is-email?` in `metabase.util`

File: src/metabase/sync_database/analyze.clj [hunk 2]
```diff
@@ -134,6 +134,45 @@
...
+        ;; If val is non-nil, check that it's a JSON dictionary or array. We don't want to mark Fields containing other
+        ;; types of valid JSON values as :json (e.g. a string representation of a number or boolean)
+        :else                  (do (assert (valid-email? val))
+                                   (recur true more))))
+    (catch Throwable _
+      false)))
+
+(defn- test:email-special-type
+  "Mark FIELD as `:json` if it's textual, doesn't already have a special type, the majority of it's values are non-nil, and all of its non-nil values
+   are valid serialized JSON dictionaries or arrays."
```
Comment (camsaul) [inline src/metabase/sync_database/analyze.clj]: copypasta?

File: test/metabase/sync_database/analyze_test.clj [hunk 3]
```diff
@@ -60,3 +60,12 @@
...
 (expect false (values-are-valid-json? ["false"]))
+
+;; Check that things that are valid emails are marked as Emails
+(expect true (values-are-valid-email? ["helper@metabase.com"]))
+(expect true (values-are-valid-email? ["helper@metabase.com", "someone@here.com", "help@nope.com"]))
+
+(expect false (values-are-valid-email? ["\"A string should not cause a Field to be marked as email\""]))
+(expect false (values-are-valid-email? [100]))
+(expect false (values-are-valid-email? ["true"]))
+(expect false (values-are-valid-email? ["false"]))
```
Comment (camsaul) [inline test/metabase/sync_database/analyze_test.clj]: how about a test with some valid & some invalid

and how about a test with some `nil` values and valid values

**Comment by camsaul** [comment]
@salsakran I see the new analysis functions you added but I don't see them being called anywhere 😕

File: src/metabase/sync_database/analyze.clj [hunk 4]
```diff
@@ -134,6 +134,45 @@
...
+  [driver field field-stats]
+  (if-not (and (not (:special_type field))
+               (not (isa? (:base_type field) :type/Text)))
```
Comment (camsaul) [inline src/metabase/sync_database/analyze.clj]: I guess this logic is the same as the function you copied it from but it looks confusing and wrong :frowning_face: 

I think we should this one and the place you copied it from to be:

```clojure
(if (or (:special_type field)
        (not (isa? (:base_type field) :type/Text)))
  field-stats
  ;; ...
  )
```

The logic should be "check for values if it doesn't already have a special type and is textual"
but right now the logic looks like "check for values if it doesn't already have a special type and is *not* textual

File: src/metabase/sync_database/analyze.clj [hunk 5]
```diff
@@ -134,6 +134,45 @@
...
+    ;; this field isn't suited for this test
+    field-stats
+    ;; check for json values
```
Comment (camsaul) [inline src/metabase/sync_database/analyze.clj]: 🍝

File: src/metabase/sync_database/analyze.clj (hunk 4 above)
Comment (salsakran) [inline src/metabase/sync_database/analyze.clj]: Yeah, I didn't really know why that was being checked.

Comment (camsaul) [inline src/metabase/sync_database/analyze.clj]: me either. I didn't write that part 😉

**Comment by salsakran** [comment]
@camsaul think I fixed all the mistakes you caught.

File: src/metabase/sync_database/analyze.clj [hunk 6]
```diff
@@ -134,14 +134,48 @@
...
 (defn- test:new-field
   "Do the various tests that should only be done for a new `Field`.
    We only run most of the field analysis work when the field is NEW in order to favor performance of the sync process."
   [driver field field-stats]
   (->> field-stats
        (test:no-preview-display driver field)
        (test:url-special-type   driver field)
-       (test:json-special-type  driver field)))
+       (test:json-special-type  driver field)
+       (test:email-special-type  driver field)))
```
Comment (camsaul) [inline src/metabase/sync_database/analyze.clj]: nitpicky but if things are aligned might as well keep them aligned correctly

**Comment by camsaul** [comment]
ok @salsakran the backend stuff looks good to me 👍

**Comment by salsakran** [comment]
cool, I'm a 👍 on the stuff I didn't write, so @tlrobinson merge at your leisure.

**Pull request merged**